File: backend/src/main/java/com/library/service/interfaces/BorrowingRequestService.java
package com.library.service.interfaces;

import com.library.dto.request.BorrowingRequestApprovalRequest;
import com.library.dto.request.BorrowingRequestCreationRequest;
import com.library.dto.response.BorrowingRequestResponse;

import java.util.List;

public interface BorrowingRequestService {
    BorrowingRequestResponse createRequest(BorrowingRequestCreationRequest request);
    List<BorrowingRequestResponse> getPendingRequests();
    List<BorrowingRequestResponse> getApprovedRequests();
    List<BorrowingRequestResponse> getHistoryRequests();
    List<BorrowingRequestResponse> getRequestsByMember(Long memberId);
    BorrowingRequestResponse approveRequest(Long requestId, BorrowingRequestApprovalRequest approvalRequest);
    BorrowingRequestResponse rejectRequest(Long requestId, String reason);
    BorrowingRequestResponse cancelRequest(Long requestId, String username);
    // Giao sách thực tế: tạo Borrowing, chuyển yêu cầu → COMPLETED
    BorrowingRequestResponse issueBook(Long requestId);
    // Hủy yêu cầu đã duyệt (không đến lấy): trả sách về AVAILABLE, chuyển → EXPIRED
    BorrowingRequestResponse expireRequest(Long requestId);
}

File: frontend/src/app/features/borrowings/services/borrowing-request.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  BorrowingRequestResponse,
  BorrowingRequestCreationRequest,
  BorrowingRequestApprovalRequest
} from '../models/borrowing-request.model';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class BorrowingRequestService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/borrowing-requests`;

  createRequest(request: BorrowingRequestCreationRequest): Observable<BorrowingRequestResponse> {
    return this.http.post<BorrowingRequestResponse>(this.apiUrl, request);
  }

  getPendingRequests(): Observable<BorrowingRequestResponse[]> {
    return this.http.get<BorrowingRequestResponse[]>(`${this.apiUrl}/pending`);
  }

  // Lấy danh sách yêu cầu đã duyệt - đang chờ độc giả đến lấy sách
  getApprovedRequests(): Observable<BorrowingRequestResponse[]> {
    return this.http.get<BorrowingRequestResponse[]>(`${this.apiUrl}/approved`);
  }

  getHistoryRequests(): Observable<BorrowingRequestResponse[]> {
    return this.http.get<BorrowingRequestResponse[]>(`${this.apiUrl}/history`);
  }

  getRequestsByMember(memberId: number): Observable<BorrowingRequestResponse[]> {
    return this.http.get<BorrowingRequestResponse[]>(`${this.apiUrl}/member/${memberId}`);
  }

  approveRequest(id: number): Observable<BorrowingRequestResponse> {
    return this.http.post<BorrowingRequestResponse>(`${this.apiUrl}/${id}/approve`, {});
  }

  issueBook(id: number): Observable<BorrowingRequestResponse> {
    return this.http.post<BorrowingRequestResponse>(`${this.apiUrl}/${id}/issue`, {});
  }

  // Hủy yêu cầu đã duyệt vì độc giả không đến lấy
  expireRequest(id: number): Observable<BorrowingRequestResponse> {
    return this.http.post<BorrowingRequestResponse>(`${this.apiUrl}/${id}/expire`, {});
  }

  rejectRequest(id: number, reason: string): Observable<BorrowingRequestResponse> {
    return this.http.post<BorrowingRequestResponse>(`${this.apiUrl}/${id}/reject?reason=${encodeURIComponent(reason)}`, {});
  }

  cancelRequest(id: number): Observable<BorrowingRequestResponse> {
    return this.http.put<BorrowingRequestResponse>(`${this.apiUrl}/${id}/cancel`, {});
  }
}
